core/base/range.hpp

```diff
@@ -0,0 +1,966 @@
+
+// binary bitwise
+GKO_DEFINE_SIMPLE_BINARY_OPERATION(bitwise_or, first | second);
+GKO_DEFINE_SIMPLE_BINARY_OPERATION(bitwise_and, first &second);
+GKO_DEFINE_SIMPLE_BINARY_OPERATION(bitwise_xor, first ^ second);
+GKO_DEFINE_SIMPLE_BINARY_OPERATION(left_shift, first << second);
+GKO_DEFINE_SIMPLE_BINARY_OPERATION(right_shift, first >> second);
+
+// common binary functions
+GKO_DEFINE_SIMPLE_BINARY_OPERATION(max_operation, max(first, second));
+GKO_DEFINE_SIMPLE_BINARY_OPERATION(min_operation, min(first, second));
+
+}  // namespace detail
+}  // namespace accessor
+
+
+// binary arithmetic
+GKO_ENABLE_BINARY_RANGE_OPERATION(add, operator+, accessor::detail::add);
+GKO_ENABLE_BINARY_RANGE_OPERATION(sub, operator-, accessor::detail::sub);
+GKO_ENABLE_BINARY_RANGE_OPERATION(mul, operator*, accessor::detail::mul);
+GKO_ENABLE_BINARY_RANGE_OPERATION(div, operator/, accessor::detail::div);
+GKO_ENABLE_BINARY_RANGE_OPERATION(mod, operator%, accessor::detail::mod);
+
+// relational
+GKO_ENABLE_BINARY_RANGE_OPERATION(less, operator<, accessor::detail::less);
+GKO_ENABLE_BINARY_RANGE_OPERATION(greater, operator>,
+                                  accessor::detail::greater);
+GKO_ENABLE_BINARY_RANGE_OPERATION(less_or_equal, operator<=,
+                                  accessor::detail::less_or_equal);
+GKO_ENABLE_BINARY_RANGE_OPERATION(greater_or_equal, operator>=,
+                                  accessor::detail::greater_or_equal);
+GKO_ENABLE_BINARY_RANGE_OPERATION(equal, operator==, accessor::detail::equal);
+GKO_ENABLE_BINARY_RANGE_OPERATION(not_equal, operator!=,
+                                  accessor::detail::not_equal);
+
+// binary logical
+GKO_ENABLE_BINARY_RANGE_OPERATION(logical_or, operator||,
+                                  accessor::detail::logical_or);
+GKO_ENABLE_BINARY_RANGE_OPERATION(logical_and, operator&&,
+                                  accessor::detail::logical_and);
+
+// binary bitwise
+GKO_ENABLE_BINARY_RANGE_OPERATION(bitwise_or, operator|,
+                                  accessor::detail::bitwise_or);
+GKO_ENABLE_BINARY_RANGE_OPERATION(bitwise_and, operator&,
+                                  accessor::detail::bitwise_and);
+GKO_ENABLE_BINARY_RANGE_OPERATION(bitwise_xor, operator^,
+                                  accessor::detail::bitwise_xor);
+GKO_ENABLE_BINARY_RANGE_OPERATION(left_shift, operator<<,
+                                  accessor::detail::left_shift);
+GKO_ENABLE_BINARY_RANGE_OPERATION(right_shift, operator>>,
+                                  accessor::detail::right_shift);
+
+// common binary functions
+GKO_ENABLE_BINARY_RANGE_OPERATION(max_operaton, max,
+                                  accessor::detail::max_operation);
+GKO_ENABLE_BINARY_RANGE_OPERATION(min_operaton, min,
+                                  accessor::detail::min_operation);
+
+
+// special binary range functions
+namespace accessor {
+
+
+template <gko::detail::operation_kind Kind, typename FirstAccessor,
```

Comment:
Well, somewhere in between. You are supposed to know what its type means for debugging purposes - that it is a representation of the `mmul` operation, which you can figure out just by looking at the name of the class, but everything else is implementation detail. You are never supposed to create them directly, or call any of their methods.  There are actually a lot more of them in the same namespace defined by the macros. All of them are in fact "accessors" as much as `row_major` is, so that's why they are directly in the `accessor` namespace.
Not sure if putting them in the `accessor::detail` would do much good - we would just get more incomprehensible error messages. Sometimes its a bit hard to figure out what should and what shouldn't be in `detail`.

Reply:
The whole implementation of the ranges is a bit difficult to document, as all the top-level stuff that the user actually calls is automatically generated (since most of it is boilerplate, with a 1-line operation hidden underneath it). There's no good place to put per-function, or per-class documentation. If you put it into any of the non-boilerplate parts, than you're telling the user that they are supposed to use it, and this is not true.

So my solution when writing documentation for ranges was to not even try doing it, and just put everything related to range operations into the `range` class description. That's everything the user will ever need.

One more thing about `detail` namespaces: I didn't start using it to hide something from the user or the documentation, I just used it to avoid unnecessary namespace pollution. Mostly these are template things that in a perfect world wouldn't even be named entities, but they have to be due to language restrictions (e.g. you cannot write a compile-time for loop without introducing new names somewhere, even though a for loop shouldn't have a name). `mmul_operation` is not something that the user should care about, but it is something that should have a name, so that's why it's not in `detail`. We can also start an issue about what does go into `detail`, and what does not. This is something we haven't discussed yet, and I cannot say I'm completely consistent with it, mostly I went with what "feels right".